File: crates/brainwires-inference/examples/agent_pool.rs
//! Example: Concurrent agent management with AgentPool
//!
//! Shows how to set up shared infrastructure (CommunicationHub, FileLockManager),
//! spawn multiple agents into an `AgentPool`, monitor their status, and collect
//! results. Uses a simple `MockProvider` that returns "Done" immediately.
//!
//! Run: cargo run -p brainwires-inference --example agent_pool

use std::sync::Arc;

use anyhow::Result;
use async_trait::async_trait;
use futures::stream::BoxStream;

use brainwires_agent::brainwires_core::{
    ChatOptions, ChatResponse, Message, Provider, StreamChunk, Task, Tool, ToolContext, ToolResult,
    ToolUse, Usage,
};
use brainwires_agent::brainwires_tool_runtime::ToolExecutor;
use brainwires_agent::{AgentMessage, CommunicationHub, FileLockManager, LockType};
use brainwires_inference::{AgentPool, TaskAgentConfig};

// ── Mock Provider ──────────────────────────────────────────────────────────

/// A minimal provider that always returns a fixed "Done" response.
struct MockProvider;

#[async_trait]
impl Provider for MockProvider {
    fn name(&self) -> &str {
        "mock"
    }

    async fn chat(
        &self,
        _messages: &[Message],
        _tools: Option<&[Tool]>,
        _options: &ChatOptions,
    ) -> Result<ChatResponse> {
        Ok(ChatResponse {
            message: Message::assistant("Done"),
            finish_reason: Some("stop".to_string()),
            usage: Usage::default(),
        })
    }

    fn stream_chat<'a>(
        &'a self,
        _messages: &'a [Message],
        _tools: Option<&'a [Tool]>,
        _options: &'a ChatOptions,
    ) -> BoxStream<'a, Result<StreamChunk>> {
        Box::pin(futures::stream::empty())
    }
}

// ── No-Op Tool Executor ───────────────────────────────────────────────────

/// A tool executor that returns "ok" for any tool call.
struct NoOpExecutor;

#[async_trait]
impl ToolExecutor for NoOpExecutor {
    async fn execute(&self, tu: &ToolUse, _ctx: &ToolContext) -> Result<ToolResult> {
        Ok(ToolResult::success(tu.id.clone(), "ok".to_string()))
    }

    fn available_tools(&self) -> Vec<Tool> {
        vec![]
    }
}

// ── Main ──────────────────────────────────────────────────────────────────

#[tokio::main]
async fn main() -> Result<()> {
    tracing_subscriber::fmt::init();

    // 1. Create shared infrastructure
    let hub = Arc::new(CommunicationHub::new());
    let lock_manager = Arc::new(FileLockManager::new());

    println!("=== Agent Pool Demo ===\n");

    // 2. Create the agent pool (max 5 concurrent agents)
    let pool = AgentPool::new(
        5,
        Arc::new(MockProvider),
        Arc::new(NoOpExecutor),
        Arc::clone(&hub),
        Arc::clone(&lock_manager),
        "/tmp/demo-project",
    );

    // 3. Spawn multiple agents with different tasks
    let make_config = || {
        let mut cfg = TaskAgentConfig::default();
        cfg.validation_config = None;
        Some(cfg)
    };

    let id1 = pool
        .spawn_agent(
            Task::new("task-1", "Implement authentication module"),
            make_config(),
        )
        .await?;
    println!("Spawned agent: {id1}");

    let id2 = pool
        .spawn_agent(
            Task::new("task-2", "Add unit tests for parser"),
            make_config(),
        )
        .await?;
    println!("Spawned agent: {id2}");

    let id3 = pool
        .spawn_agent(
            Task::new("task-3", "Refactor error handling"),
            make_config(),
        )
        .await?;
    println!("Spawned agent: {id3}");

    // 4. Print pool stats
    let stats = pool.stats().await;
    println!("\nPool stats:");
    println!("  Max agents:   {}", stats.max_agents);
    println!("  Total agents: {}", stats.total_agents);
    println!("  Running:      {}", stats.running);
    println!("  Completed:    {}", stats.completed);

    // 5. List active agents
    let active = pool.list_active().await;
    println!("\nActive agents:");
    for (id, status) in &active {
        println!("  {}: {}", &id[..20.min(id.len())], status);
    }

    // 6. Await all agents and print results
    println!("\nAwaiting all agent completions...");
    let results = pool.await_all().await;
    for (id, result) in &results {
        match result {
            Ok(r) => println!(
                "  {} => success={}, iterations={}, summary={}",
                &id[..20.min(id.len())],
                r.success,
                r.iterations,
                &r.summary[..50.min(r.summary.len())]
            ),
            Err(e) => println!("  {} => error: {}", &id[..20.min(id.len())], e),
        }
    }

    // 7. Demonstrate file lock coordination
    println!("\n=== File Lock Coordination ===");

    // Multiple concurrent reads are allowed
    {
        let read1 = lock_manager
            .acquire_lock("demo-agent-1", "src/lib.rs", LockType::Read)
            .await?;
        let read2 = lock_manager
            .acquire_lock("demo-agent-2", "src/lib.rs", LockType::Read)
            .await?;
        println!("Two agents reading src/lib.rs concurrently - OK");

        // Explicitly release before the guards drop (Drop spawns async tasks
        // that race with the next acquire_lock call). Forget the guards to
        // prevent the Drop handler from double-releasing.
        lock_manager
            .release_lock("demo-agent-1", "src/lib.rs", LockType::Read)
            .await?;
        lock_manager
            .release_lock("demo-agent-2", "src/lib.rs", LockType::Read)
            .await?;
        std::mem::forget(read1);
        std::mem::forget(read2);
    }

    // Exclusive write lock
    {
        let _write = lock_manager
            .acquire_lock("demo-agent-1", "src/lib.rs", LockType::Write)
            .await?;
        println!("One agent has exclusive write access to src/lib.rs - OK");
    }

    // 8. Demonstrate hub message broadcasting
    println!("\n=== Communication Hub ===");

    hub.register_agent("listener-1".to_string()).await?;
    hub.register_agent("listener-2".to_string()).await?;

    hub.broadcast(
        "orchestrator".to_string(),
        AgentMessage::StatusUpdate {
            agent_id: "orchestrator".to_string(),
            status: "planning".to_string(),
            details: Some("Starting cycle 1".to_string()),
        },
    )
    .await?;

    // Each registered agent receives the broadcast
    for listener in &["listener-1", "listener-2"] {
        match hub.receive_message(listener).await {
            Some(env) => println!("  {listener} received: {:?}", env.message),
            None => println!("  {listener}: no messages"),
        }
    }

    println!("\nAgent pool demo complete.");
    Ok(())
}

File: deno/examples/agents/agent_pool.ts
// Example: Agent Pool Management
// Demonstrates AgentPool for managing multiple concurrent TaskAgents,
// monitoring status, and collecting results. Uses AgentPoolStats
// to track pool health. Also shows CommunicationHub and FileLockManager.
// Run: deno run deno/examples/agents/agent_pool.ts

import {
  type AgentPoolStats,
  CommunicationHub,
  FileLockManager,
} from "@rullama/agent";
import { Task } from "@rullama/core";

async function main() {
  console.log("=== Agent Pool Demo ===\n");

  // NOTE: AgentPool requires a Provider and AgentContext to actually spawn
  // agents. This example demonstrates the pool stats, communication hub,
  // and file lock manager APIs that don't require a live provider.

  // 1. Demonstrate CommunicationHub
  console.log("--- 1. Communication Hub ---");

  const hub = new CommunicationHub();
  hub.registerAgent("agent-1");
  hub.registerAgent("agent-2");
  hub.registerAgent("listener-1");

  console.log(`  Registered agents: ${hub.agentCount()}`);

  // Send a direct message
  hub.sendMessage("agent-1", "agent-2", {
    type: "status_update",
    agentId: "agent-1",
    status: "working",
    details: "Implementing auth module",
  });

  // Broadcast to all
  hub.broadcast("orchestrator", {
    type: "status_update",
    agentId: "orchestrator",
    status: "planning",
    details: "Starting cycle 1",
  });

  // Check messages (receiveMessage returns a Promise)
  const msgPromise = hub.receiveMessage("agent-2");
  if (msgPromise) {
    const msg = await msgPromise;
    console.log(`  agent-2 received: ${msg.message.type} from ${msg.from}`);
  }

  const broadcastPromise = hub.receiveMessage("listener-1");
  if (broadcastPromise) {
    const broadcastMsg = await broadcastPromise;
    console.log(
      `  listener-1 received broadcast: ${broadcastMsg.message.type} from ${broadcastMsg.from}`,
    );
  }

  // 2. Demonstrate FileLockManager
  console.log("\n--- 2. File Lock Manager ---");

  const lockManager = new FileLockManager();

  // Multiple concurrent reads are allowed
  const read1 = lockManager.acquireLock("agent-1", "src/lib.rs", "read");
  const read2 = lockManager.acquireLock("agent-2", "src/lib.rs", "read");
  console.log("  Two agents reading src/lib.rs concurrently - OK");

  // Release reads via the guard
  read1.release();
  read2.release();

  // Exclusive write lock
  const writeGuard = lockManager.acquireLock("agent-1", "src/lib.rs", "write");
  console.log("  agent-1 has exclusive write access to src/lib.rs - OK");

  // Attempt conflicting read while write is held
  let conflictDetected = false;
  try {
    lockManager.acquireLock("agent-2", "src/lib.rs", "read");
  } catch {
    conflictDetected = true;
  }
  console.log(
    `  agent-2 read while write held: ${
      conflictDetected ? "BLOCKED (expected)" : "OK"
    }`,
  );

  writeGuard.release();

  // Lock stats
  const lockStats = lockManager.stats();
  console.log("\n  Lock stats:");
  console.log(`    Files with locks: ${lockStats.totalFiles}`);
  console.log(`    Write locks: ${lockStats.totalWriteLocks}`);
  console.log(`    Read locks: ${lockStats.totalReadLocks}`);

  // 3. Demonstrate Task creation for pool usage
  console.log("\n--- 3. Task Creation ---");

  const tasks = [
    new Task("task-1", "Implement authentication module"),
    new Task("task-2", "Add unit tests for parser"),
    new Task("task-3", "Refactor error handling"),
  ];

  for (const task of tasks) {
    console.log(
      `  Created: ${task.id} - "${task.description}" [${task.status}]`,
    );
  }

  // 4. Show what AgentPoolStats looks like
  console.log("\n--- 4. Pool Stats Structure ---");

  const mockStats: AgentPoolStats = {
    maxAgents: 5,
    totalAgents: 3,
    running: 2,
    completed: 1,
    failed: 0,
  };

  console.log(`  Max agents:   ${mockStats.maxAgents}`);
  console.log(`  Total agents: ${mockStats.totalAgents}`);
  console.log(`  Running:      ${mockStats.running}`);
  console.log(`  Completed:    ${mockStats.completed}`);
  console.log(`  Failed:       ${mockStats.failed}`);

  // 5. Hub message types showcase
  console.log("\n--- 5. Hub Message Types ---");

  hub.registerAgent("worker-1");

  hub.sendMessage("orchestrator", "worker-1", {
    type: "task_request",
    taskId: "task-1",
    description: "Implement authentication module",
    priority: 3,
  });

  hub.sendMessage("worker-1", "agent-1", {
    type: "task_result",
    taskId: "task-1",
    success: true,
    result: "Auth module implemented with JWT support",
  });

  hub.sendMessage("agent-1", "agent-2", {
    type: "help_request",
    requestId: "help-1",
    topic: "error handling",
    details: "Need advice on error propagation pattern",
  });

  const workerMsg = hub.tryReceiveMessage("worker-1");
  if (workerMsg) {
    console.log(`  worker-1 received: ${workerMsg.message.type}`);
  }

  const agent1Msg = hub.tryReceiveMessage("agent-1");
  if (agent1Msg) {
    console.log(`  agent-1 received: ${agent1Msg.message.type}`);
  }

  const agent2Msg = hub.tryReceiveMessage("agent-2");
  if (agent2Msg) {
    console.log(`  agent-2 received: ${agent2Msg.message.type}`);
  }

  console.log(`\n  Registered agents: [${hub.listAgents().join(", ")}]`);

  console.log("\nAgent pool demo complete.");
}

await main();
